[MRG] Memory check based on memory profiler

aabadie opened the pull request:
This should fix #736 but it's not fully functional. The problem is memory_profiler cannot be really trusted when a function increase the memory less than 1MiB. I also couldn't apply it on a real test.

Here is an example (but it doesn't work) with it `test_masking.py`:
```python

from nilearn._utils.testing import with_memory_usage, assert_memory_less_than

@with_memory_usage
def test_memory_usage_with_multi_epi_mask():
    # As it calls intersect_masks, we only test resampling here.
    # Same masks as test_intersect_masks
    mask_a = np.zeros((400, 400, 1), dtype=np.bool)
    mask_a[200:400, 200:400] = 1
    mask_a_img = Nifti1Image(mask_a.astype(int), np.eye(4))

    mask_b = np.zeros((800, 800, 1), dtype=np.bool)
    mask_b[200:600, 200:600] = 1
    mask_b_img = Nifti1Image(mask_b.astype(int), np.eye(4) / 2.)

    mask_ab = np.zeros((400, 400, 1), dtype=np.bool)
    mask_ab[200, 200] = 1

   # arbitrary ground truth value...
    test_mem = 2 * (mask_b_img.get_data().nbytes +
                    mask_b_img.get_data().nbytes +
                    mask_a_img.get_data().nbytes) / (1024 ** 2)
    assert_memory_less_than(test_mem, 0.1,
                            compute_multi_epi_mask,
                            [mask_a_img, mask_b_img], threshold=1., opening=0,
                            target_affine=np.eye(4),
                            target_shape=(400, 400, 1))
```

Maybe an expert such as @AlexandreAbraham will have a better idea :)

aabadie commented:
For the record, I pushed a test function that checks that the `_iter_check_niimg` doesn't do any memory copy (as suggested by @AlexandreAbraham). The memory usage is as expected.

AlexandreAbraham commented:
That's great !

I have another suggestion (not for your PR, in general). In nilearn, we often use weird coding patterns (I'm thinking of taking the sum of a numpy array to know if there is nan inside for example). It would be interesting to keep a performance test (time and memory usage) for some of these patterns.

It is useful for other developers and, for example, we could bench the Nifti1Image format, on which we rely heavily, and thus spot if there is performance regression.

Benching the I/O is also interesting for cluster computation: some design patterns may not be favorable on a cluster because of longer file access time.

This is not a trivial work. What do you think?

aabadie commented:
> I have another suggestion (not for your PR, in general). In nilearn, we often
> use weird coding patterns (I'm thinking of taking the sum of a numpy array
> to know if there is nan inside for example). It would be interesting to keep
> a performance test (time and memory usage) for some of these patterns.

For computation time testing, it's difficult because it depends on the speed of the cpu running the test. Using reference measures computed from the current master and stored in a dedicated file, we could compare the measures of a PR against them and raise an error in case of a too large difference.

> Benching the I/O is also interesting for cluster computation: some design
> patterns may not be favorable on a cluster because of longer file access
> time.

I see the point. I don't how one can check the underlying I/O system calls from Python.
Maybe this could be done using a call to an external command like (only available on posix):
```bash
# Return the number of system calls to read
strace python -c "some python code with I/O" | grep read | wc -l
# Return the number of system calls to write
strace python -c "some python code with I/O" | grep write | wc -l
```
The output of the commands above could then be sent back to and analyzed be the global python test script.

> This is not a trivial work.

I agree, but it's not impossible.

lesteve commented:
>>  This is not a trivial work.

> I agree, but it's not impossible.

scikit-learn tried to do these kind of performance monitoring at one point during a GSoC. From what I have heard, a significant amount of time was invested with little return because this is a lot of work to maintain the infrastructure to do it. From what I remember it was based on vbench which ended up changing its internals and breaking the work that was done on scikit-learn.

lesteve commented:
You seem to have introduced a PEP8 (line too long). You can look at the Travis build output [here](https://travis-ci.org/nilearn/nilearn/jobs/104104951). The best is if you run ./continuous_integration/flake8_diff.sh in your local code.

aabadie commented:
> You seem to have introduced a PEP8 (line too long)

Indeed, I changed the configuration of my IDE flake8 to accept lines up to 80 characters...

lesteve commented on nilearn/_utils/testing.py:
```diff
@@ -60,6 +61,43 @@ def assert_warns(warning_class, func, *args, **kw):
+                            callable_obj=None, *args, **kwargs):
+    if expected_value < 1:  # MiB
+        raise ValueError("Memory profiler results cannot be trusted for "
+                         "values less than 1MiB (1e6).")
```
I am not sure what 1e6 means and I think it is misleading since it actually means 1024 ** 2 bytes.

lesteve commented on nilearn/_utils/testing.py:
```diff
@@ -60,6 +61,43 @@ def assert_warns(warning_class, func, *args, **kw):
+
+
+def assert_memory_less_than(expected_value, tolerance,
+                            callable_obj=None, *args, **kwargs):
```
What's the point of having callable_obj=None by default?

aabadie replied:
Probably the same as the one I took as example here: https://github.com/nilearn/nilearn/pull/959/files#diff-84deb1797526e96db1bf5f1c72f508eaR36
I'll remove the default parameter.

lesteve commented on nilearn/_utils/testing.py:
```diff
@@ -60,6 +61,43 @@ def assert_warns(warning_class, func, *args, **kw):
+        return dummy_func
+
+    def memory_used(func, *args, **kwargs):
+        """Return an invalid value."""
```
why not do `memory_used = None` like you do for memory_usage?

aabadie replied:
Indeed this is not necessary.

lesteve commented on nilearn/_utils/testing.py:
```diff
@@ -60,6 +61,43 @@ def assert_warns(warning_class, func, *args, **kw):
+
+def assert_memory_less_than(expected_value, tolerance,
+                            callable_obj=None, *args, **kwargs):
+    if expected_value < 1:  # MiB
```
I thought we agreed to check the used memory as measured by memory_profiler rather that the expected_value. Also from our little experiments I would say that setting the threshold at 10 MiB is more reasonable.

aabadie replied:
> I thought we agreed to check the used memory as measured by memory_profiler rather that the expected_value.

I think that we didn't converge. Thinking of it, `expected_value` should be renamed to  `memory_limit` to be more in the spirit of the initial desired behaviour.

> I would say that setting the threshold at 10 MiB is more reasonable.

Agreed, will change.

lesteve replied:
> I think that we didn't converge

I should work on my impression of converging then :-). Given the %memit behaviour we saw, I would feel a lot more confident to check that the memory usage is not smaller than 10 MiB. You may need to change your test to make sure that the function uses some memory for example by creating the array inside the function if you want to make sure that something doesn't make any copy, e.g. in your test_iter_niimg stuff but that sounds fine to me.

aabadie replied:
> Given the %memit behaviour we saw, I would feel a lot more confident to check that the memory usage is not smaller than 10 MiB

I added an extra check one the memory is measured. You were right : with 10MiB objects, the measured memory was 0.0... It works as expected with larger objects.

GaelVaroquaux commented:
> From what I remember it was based on vbench which ended up changing its
> internals and breaking the work that was done on scikit-learn.

Pandas is now using something different, and they are very happy with it.
We should look at that. But nilearn is probably not the place to set this
up; rather a bigger project, with more manpowerful.

aabadie commented:
> But nilearn is probably not the place to set this up; rather a bigger project, with more manpowerful.

scikit-learn ?

aabadie commented:
> Pandas is now using something different, and they are very happy with it.

Looking at their repos, it seems they are still using vbench. But I may have missed something.

GaelVaroquaux commented:
> scikit-learn ?

Yes. But let's first discuss this in an engineering meeting at the lab.
Do you want to schedule one? (doodle?)

aabadie commented:
> Do you want to schedule one? (doodle?)

Sure. Let's do it next wednesday afternoon or thursday.

lesteve commented:
You have some Travis failures.

lesteve commented on circle.yml:
```diff
@@ -17,6 +17,7 @@ dependencies:
     - pip install scipy
     - pip install scikit-learn
     - pip install nose-timer
+    - pip install memory_profiler psutil
```
Maybe we don't really care about running memory_profiler tests on CircleCI, given that we are already running them on Travis.

aabadie replied:
Yeah, I just wanted to compare the results across CI's (as the tests fail on travis). I'll revert this.

aabadie commented:
> You have some Travis failures.

Yes this was because my memory_profiler confidence limit was too low (10MiB), I increased it to 100MiB.
I also removed the check on the expected memory_limit and refactored a bit the `assert_memory_less_than` function.

Waiting for CI now.

lesteve commented on nilearn/_utils/testing.py:
```diff
@@ -60,6 +61,47 @@ def assert_warns(warning_class, func, *args, **kw):
+        """Compute memory usage when executing func."""
+        gc.collect()
+        mem_use = memory_usage((func, args, kwargs),
+                               interval=.001, timeout=1)
```
Quickly looking at the memory_profiler code, it looks like timeout does nothing when memory usage is called with a tuple (func, args, kwargs).

aabadie replied:
I removed it.

lesteve commented:
My 2c: I wouldn't spend too much time making it work on AppVeyor, especially if that involves sprinkling the tested functions with time.sleep ...

Having tests on Travis is already an improvement on what we currently have.

aabadie commented:
>  Having tests on Travis is already an improvement on what we currently have.

Agreed. I tried to better understand what happens on windows. It seems that memory_profiler measures are less stable on this OS. My last attempt fixes this but, as you said, it introduces not so nice workarounds.

I'll remove the memory_profiler installation from appveyor.

GaelVaroquaux commented:
> My 2c: I wouldn't spend too much time making it work on AppVeyor,
> especially if that involves sprinkling the tested functions with
> time.sleep ...

> Having tests on Travis is already an improvement on what we currently have.

Absolutely. If we have memory regressions, they are most likely not
specific to Windows.

aabadie commented:
Just for the record, I think this is PR is ready for a re-review (nearly mergeable).

lesteve commented on nilearn/tests/test_niimg_conversions.py:
```diff
@@ -341,6 +343,23 @@ def test_iter_check_niimgs():
+    # measures with memory_profiler.
+    mem_offset = b'a' * 100 * 1024 ** 2
+    list(_iter_check_niimg(list_img_3d))
+    del mem_offset
```
Is this still necessary? I seem to remember it was added when you tried to make the AppVeyor tests pass.

aabadie replied:
Codacy disliked the previous code, if I remove the `del` line there's a flake8 issue on travis.

lesteve replied:
FWIW I don't think codacy brings much to the table. I'll probably remove it at one point. You can use `return mem_offset` if you really care about getting rid of the flake8 warning.

aabadie replied:
I updated this. Travis and Codacy should be happy now.

aabadie commented on nilearn/tests/test_testing.py:
```diff
@@ -1,10 +1,36 @@
+
+def create_object(size):
+    """Just create and return an object containing `size` bytes."""
+    mem_use = b'a' * size
```
This generates a flake8 issue on travis (non used variable). Replacing by `b'a' * size` makes codacy fails...

lesteve replied:
Use `return mem_use`

lesteve commented:
LGTM, let's merge this one and revisit potentially if we think we can improve it.
